### PR TITLE
Add streamlit webui

### PR DESCRIPTION
**Pull request by Ryu1845**

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
+# AGPL: a notification must be added stating that changes have been made to that file.
+
```
Comment (pepinu) [inline app.py:2]: don't you want to make it class based? take a look at the following proposal and see if you could get inspired by it 

```

class StreamlitUI:
    def __init__(self):
        # all ui creation done here

class App:
    def __init__(self):
        self.tts = load_model()
        self.ui = StreamlitUI()
        ...
    
    def run(self):
        if not st.button("Start"):
            return
        # if button was clicked, process the input
        # prepare 
        # run inference
            ...


def main():
    app = App()
    app.run()

if __name__ == "__main__":
     main()

**Review by pepinu — CHANGES_REQUESTED** [review]
great effort, however this is very inefficient and hard to read.
model should be initialised once per session, otherwise the model will be created and loaded every time you click process button. this can be done with session state and/or st.cache:
https://docs.streamlit.io/library/api-reference/session-state
https://docs.streamlit.io/library/advanced-features/caching

there could be an argument of not closing the model between sessions and making gpu resource busy but I don't think this is for any other than demo purposes

File: app.py (hunk 1 above)
Comment (Ryu1845) [reply app.py:2]: I'm not really a fan of classes when they don't reduce the amount of code

File: app.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                "Produce Debug State",
+                help="Whether or not to produce debug_state.pth, which can aid in reproducing problems. Defaults to true.",
+                value=True,
+            )
+
+
+    if st.button("Start"):
+        with st.spinner(f"Generating {candidates} candidates for voice {voice} (seed={seed}). You can see progress in the terminal"):
+            os.makedirs(output_path, exist_ok=True)
+            tts = TextToSpeech(models_dir=model_dir, high_vram=high_vram, kv_cache=kv_cache)
```
[the remark this answers is not in the document]
Comment (Ryu1845) [reply app.py:128]: Thanks, will do.

**Comment by Ryu1845** [comment]
Thanks for the review, I'll make those changes tomorrow

File: app.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ryu1845) [reply app.py:128]: Actually I don't understand the code, if tts in initialized in global scope, why even make a function?

**Comment by Ryu1845** [comment]
@152334H could you review?

File: app.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                    os.path.join(output_path, filename),
+                    g.squeeze(0).cpu(),
+                    24000,
+                )
+                audio_buffer = BytesIO()
+                torchaudio.save(
+                    audio_buffer,
+                    g.squeeze(0).cpu(),
+                    24000,
+                )
```
Comment (152334H) [inline app.py:176]: bug: missed `format="wav"` (which is included in the copy of this line below); i'll edit this

File: app.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            sampler = st.radio(
+                "Sampler",
+                SAMPLERS,
+                help="override the sampler used for diffusion (default depends on preset)",
+                index=1
+            )
+            steps = st.number_input(
+                "Steps",
+                help="Override the steps used for diffusion (default depends on preset)",
+                value=100,
```
Comment (152334H) [inline app.py:72]: really high honestly, even DDIM doesn't need to be this high, i'll lower this unless there's a good reason not to

File: app.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            half = st.checkbox(
+                "Half-Precision",
+                help="Enable autocast to half precision for autoregressive model",
+                value=False,
+            )
+            kv_cache = st.checkbox(
+                "Key-Value Cache",
+                help="Enable (partially broken) kv_cache usage, leading to drastic speedups but worse memory usage + results",
+                value=False,
+            )
```
Comment (152334H) [inline app.py:108]: i don't think this should be disabled by default but i'll leave it there until someone else protests

File: app.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+            """#### Debug"""
+            produce_debug_state = st.checkbox(
+                "Produce Debug State",
+                help="Whether or not to produce debug_state.pth, which can aid in reproducing problems. Defaults to true.",
+                value=True,
+            )
+
+    if 'tts' not in st.session_state:
+        st.session_state.tts = TextToSpeech(models_dir=model_dir, high_vram=high_vram, kv_cache=kv_cache)
+    tts = st.session_state.tts
```
Comment (152334H) [inline app.py:127]: this just doesn't work. changing the checkboxes on the browser does not actually reload the configuration settings for TextToSpeech(). this is a significant issue. i'll try to fix it too

**Review by 152334H — COMMENTED** [review]
:bug:s

File: app.py (hunk 6 above)
Comment (152334H) [reply app.py:129]: i [patched](https://github.com/152334H/tortoise-tts-fast/pull/12/commits/bffb8ddcf47cbc0b56f91c80ed1d769ff357f31e) this to reinit the TTS object every time the relevant config options are changed. this is suboptimal and a bettter approach would be to have a specific `TextToSpeech.update()` method that does the bare minimum required work to change options.

**Comment by 152334H** [comment]
@Ryu1845 @pepinu I'll merge this if there are no other comments

**Comment by Ryu1845** [comment]
Looks good to me

**Comment by FurkanGozukara** [comment]
I am getting this error any ideas why?

https://github.com/152334H/tortoise-tts-fast/issues/81

@Ryu1845

**Pull request merged**